### ISSUE TITLE
Crash on opening JPEG (glibmm-ERROR)

### ISSUE THREAD

**Issue by Thanatomanic** [Beep6581/RawTherapee#5207]
![50276032-f7564900-0451-11e9-8489-baad88a8a04c](https://user-images.githubusercontent.com/6567747/54084698-4cf51000-4334-11e9-8276-6a014fbe927c.JPG)
Opening this image from #5109 in RT gives me the following crash info (current dev on fully updated MSYS2 Win10):

```
(rawtherapee.exe:13660): glibmm-ERROR **: 12:58:14.784:
unhandled exception (type std::exception) in signal handler:
what: basic_string::substr: __pos (which is 18446744073709551615) > this->size() (which is 4)
```

I'm pretty sure this crash didn't happen before because #5109 was solved.

**Comment by Thanatomanic** [comment]
What is even weirder is that now that I've cleared the cache, the image does not even show up anymore in the file browser.

I just checked with the 5.5 Release version. No problems there.

**Comment by Thanatomanic** [comment]
The plot thickens: if I resave the image but strip the EXIF data it shows up and loads without problem. So, there must be some fallout from recent changes...

**Comment by Floessie** [comment]
@Thanatomanic Should be easy to bisect... I can help with a fix.

**Comment by Thanatomanic** [comment]
Working on it (and learning to bisect...) 😄

**Comment by Thanatomanic** [comment]
@Floessie The result of the bissection is this:

Bisecting: 0 revisions left to test after this (roughly 0 steps)
[70475f781aaddf49b7392236a38fe0bb7258a0a2] Derive lens name from 0xA434 and 0xA433 if other means fail (#5157)

Can you take a look?

**Comment by Floessie** [comment]
@Thanatomanic Came to the same conclusion.

It's [this line](https://github.com/Beep6581/RawTherapee/commit/70475f781aaddf49b7392236a38fe0bb7258a0a2#diff-d4c7b8ac138bac4d4484e473c84a6307R447). But it doesn't crash on every run and is hard to nail down.

The buffer handling in `rtexif` has always been a mess. I think we have a buffer overflow in one of the `toString()`s...

**Comment by Floessie** [comment]
@Thanatomanic I've not found the real culprit. `Tag::toString()` behaved as expected. Nevertheless, there seems to be a side effect when skipping [this `if`](https://github.com/Beep6581/RawTherapee/blob/dev/rtengine/imagedata.cc#L454).

Oh, wait: `lens.substr(lens.find(' ')).length() < 7` is buggy. `std::string::substr()`'s first parameter must be within range. Roel, could you fix that, as I'm AFK now?

**Comment by Thanatomanic** [comment]
@Floessie I'll check

**Comment by Thanatomanic** [comment]
@Floessie Your fix c240558 solved the crash allright. And is this something you were thinking of for the other fix?

```diff
diff --git a/rtengine/imagedata.cc b/rtengine/imagedata.cc
index a7d2d157f..19ff98d9a 100644
--- a/rtengine/imagedata.cc
+++ b/rtengine/imagedata.cc
@@ -451,14 +451,17 @@ FrameData::FrameData (rtexif::TagDirectory* frameRootDir_, rtexif::TagDirectory*
                         }
                     }
 
-                    if( !found || lens.substr(lens.find(' ')).length() < 7 ) {
-                        lt = mnote->findTag("LensID");
+                    std::size_t spacefind = lens.find(' ');
+                    if ( spacefind != std::string::npos ) {
+                        if( !found || lens.substr(spacefind).length() < 7 ) {
+                            lt = mnote->findTag("LensID");
 
-                        if ( lt ) {
-                            std::string ldata = lt->valueToString ();
+                            if ( lt ) {
+                                std::string ldata = lt->valueToString ();
 
-                            if (ldata.size() > 1) {
-                                lens = ldata;
+                                if (ldata.size() > 1) {
+                                    lens = ldata;
+                                }
                             }
                         }
                     }

**Comment by Floessie** [comment]
@Thanatomanic Exactly. :+1: 

I've modified your patch a bit to keep the previous behavior. Feel free to commit it, otherwise I'll do it tonight.

```patch
diff --git a/rtengine/imagedata.cc b/rtengine/imagedata.cc
index a7d2d157f..74cc0a992 100644
--- a/rtengine/imagedata.cc
+++ b/rtengine/imagedata.cc
@@ -448,10 +448,18 @@ FrameData::FrameData (rtexif::TagDirectory* frameRootDir_, rtexif::TagDirectory*
                                 found = true;
                                 lens = "Canon " + ldata;
                             }
+                        } else {
+                            found = lens_from_make_and_model();
                         }
                     }
 
-                    if( !found || lens.substr(lens.find(' ')).length() < 7 ) {
+                    const std::string::size_type first_space_pos = lens.find(' ');
+                    const std::string::size_type remaining_size =
+                        first_space_pos != std::string::npos
+                            ? lens.size() - first_space_pos
+                            : 0;
+
+                    if( !found || remaining_size < 7U ) {
                         lt = mnote->findTag("LensID");
 
                         if ( lt ) {
@@ -462,10 +470,6 @@ FrameData::FrameData (rtexif::TagDirectory* frameRootDir_, rtexif::TagDirectory*
                             }
                         }
                     }
-
-                    if (!found) {
-                        lens_from_make_and_model();
-                    }
                 } else if (!make.compare (0, 6, "PENTAX") || (!make.compare (0, 5, "RICOH") && !model.compare (0, 6, "PENTAX"))) {
                     // ISO at max value supported, check manufacturer specific
                     if (iso_speed == 65535 || iso_speed == 0) {
```

Best,
Flössie
